### PR TITLE
Fix discovery issues

### PR DESCRIPTION
**Pull request by gumb0**
Fixes https://github.com/ethereum/cpp-ethereum/issues/4773

It was not really a discovery problem, more like a problem of managing peers and in `Host` class, see comment below.

Also unnecessary mutex `x_pendingPeerConns` removed.

### CODE REVIEW

File: libp2p/Host.cpp [hunk 1]
```diff
@@ -60,930 +60,938 @@ ReputationManager::ReputationManager()
...
-	
-	if (!m_netPrefs.pin)
-	{
-		unsigned pendingCount = 0;
-		DEV_GUARDED(x_pendingNodeConns)
-			pendingCount = m_pendingPeerConns.size();
-		int openSlots = m_idealPeerCount - peerCount() - pendingCount + reqConn;
-		if (openSlots > 0)
-			for (auto p: toConnect)
-				if (p->peerType == PeerType::Optional && openSlots--)
```
Comment (gumb0) [inline libp2p/Host.cpp]: Here I guess the intention was to break the loop when we run out of `openSlots`. What was really happening is of course `openSlots` going negative when there are a lot of peers in `toConnect` available and it tried to connect to all of them at once.

Then `connect()` added all these peers to `m_pendingPeerConns` (the set of peers we are trying to connect to) and this variable affects `peerSlotsAvailable()`, so when the handshake was complete with any peer it checked `peerSlotsAvailablr()` (in `startPeerSession()`) and it always looked as no available slots and so it disconnected.

**Comment by gumb0** [comment]
Could also affect `Too many open files error` as sometimes it tried to open a lot of sockets at once.

File: libp2p/Host.cpp [hunk 2]
```diff
@@ -60,930 +60,926 @@ ReputationManager::ReputationManager()
...
+        }
+    }
+
+    for (auto p: toConnect)
+        if (p->peerType == PeerType::Required && reqConn++ < m_idealPeerCount)
+            connect(p);
+    
+    if (!m_netPrefs.pin)
+    {
+        int openSlots = m_idealPeerCount - peerCount() - m_pendingPeerConns.size() + reqConn;
```
Comment (chfast) [inline libp2p/Host.cpp]: Are there any type conversions happening here? What are the chances of overflow?

Comment (gumb0) [reply libp2p/Host.cpp]: Hmm, all the values on the right of assignment are `unsigned`
Not sure now if it's  safe...
realistically none of these values go anywhere near `max<int>()`

Comment (gumb0) [reply libp2p/Host.cpp]: I will better rewrite it without cast to `int`

**Comment by gumb0** [comment]
Thanks for the link 👌

**Pull request merged**